keras/utils/data_utils.py

```diff
@@ -282,3 +295,295 @@ def validate_file(fpath, file_hash, algorithm='auto', chunk_size=65535):
+        """Get batch at position `index`.
+
+        # Arguments
+            index: position of the batch in the Dataset.
+
+        # Returns
+            A batch
+        """
+        raise NotImplementedError
+
+    def __len__(self):
+        """Number of batch in the Dataset.
+
+        # Returns
+            The number of batch in the dataset.
+        """
+        raise NotImplementedError
+
+
+def get_index(ds, i):
+    """Quick fix for Python2, otherwise, it cannot be pickled.
+
+    # Arguments
+        ds: a Dataset object
+        i: index
+
+    # Returns
+        The value at index `i`.
+    """
+    return ds[i]
+
+
+class DatasetEnqueuer(object):
+    """Base class to enqueue inputs.
+    The task of an Enqueuer is to use parallelism to speed up the preprocessing.
+    This is done with processes or threads.
+
+    # Examples
+    ```python
+    enqueuer = DatasetEnqueuer(...)
+    enqueuer.start()
+    datas = enqueuer.get()
+    for data in datas:
+        # Use the inputs; training, evaluating, predicting.
+        # ... stop sometime.
+    enqueuer.close()
+    ```
+
+    The `enqueuer.get()` should be an infinite stream of datas.
+
+    """
+
+    @abstractmethod
+    def is_running(self):
+        raise NotImplemented
+
+    @abstractmethod
+    def start(self, workers=1, max_q_size=10):
+        """Start the handler's workers.
+
+        # Arguments
+            workers: number of worker threads
+            max_q_size: queue size (when full, threads could block on put())
+        """
+        raise NotImplemented
+
+    @abstractmethod
+    def stop(self, timeout=None):
+        """Stop running threads and wait for them to exit, if necessary.
+
+        Should be called by the same thread which called start().
+
+        # Arguments
+            timeout: maximum time to wait on thread.join()
+        """
+        raise NotImplemented
+
+    @abstractmethod
+    def get(self):
+        """Create a generator to extract data from the queue.
+
+        #Returns
+            A generator
+        """
+        raise NotImplemented
+
+
+class OrderedEnqueuer(DatasetEnqueuer):
+    """Builds a Enqueuer from a Dataset.
+
+    Used in `fit_generator`, `evaluate_generator`, `predict_generator`.
+
+    # Arguments
+        dataset: A `keras.data.dataset.Dataset` object.
+        pickle_safe: use multiprocessing if True, otherwise threading
+        scheduling: Sequential querying of datas if 'sequential', random otherwise.
+    """
+
+    def __init__(self, dataset, pickle_safe=False, scheduling='sequential'):
+        self.dataset = dataset
+        self.pickle_safe = pickle_safe
+        self.scheduling = scheduling
+        self.workers = 0
+        self.executor = None
+        self.queue = None
+        self.run_thread = None
+        self.stop_signal = None
+
+    def is_running(self):
+        return self.stop_signal is not None and not self.stop_signal.is_set()
+
+    def start(self, workers=1, max_q_size=10):
```

Comment:
The thing is that for every result in the queue, there is a thread running on a process. This would cause massive lag on smaller pc. 1 processes with 100 threads in it.